honeybadgermpc/hbavss_batch.py

```diff
@@ -0,0 +1,235 @@
+        # Decrypt
+        for k in range(self.t+1):
+            # print(self.my_id, k, len(encrypted_witnesses[self.my_id * (self.t+1) + k]))
+            shares[k], witnesses[k] = SymmetricCrypto.decrypt(
+                str(shared_key).encode("utf-8"), encrypted_witnesses[self.my_id * (self.t+1) + k])
+
+        ##################### Don't know why this doesn't work #####################
+        # # Decrypt
+        # for k in range(self.t+1):
+        #     shares[k], witnesses[k] = SymmetricCrypto.decrypt(
+        #         str(shared_key).encode("utf-8"), encrypted_witnesses[self.my_id][k])
+        ############################################################################
+
+        # verify & send all ok
+        for i in range(self.t+1):
+            if self.poly_commit.verify_eval(commitments[i], self.my_id+1, shares[i], witnesses[i]):
+                multicast(HbAVSSMessageType.OK)
+            else:
+                multicast(HbAVSSMessageType.IMPLICATE)
+
+        oks_recvd = 0
+        readys_recvd = 0
+        # if 2t+1 OK or t+1 READY -> send all ready
+        while oks_recvd < 2 * self.t + 1 and readys_recvd < self.t + 1:
+            _, avss_msg = await recv()  # First value is the `sid`
+            if avss_msg == HbAVSSMessageType.OK:
+                oks_recvd += 1
+            if avss_msg == HbAVSSMessageType.READY:
+                readys_recvd += 1
+
+        # logging.debug("[%d][HbAvssBatch] 2t+1 OKs or t+1 READYs received.", self.my_id)
+        multicast(HbAVSSMessageType.READY)
+
+        # if 2t+1 ready -> output shares
+        while readys_recvd < 2 * self.t + 1:
+            _, avss_msg = await recv()  # First value is the `sid`
+            if avss_msg == HbAVSSMessageType.READY:
+                readys_recvd += 1
+
+        # logging.debug("[%d][HbAvssBatch] 2t+1 READYs received.", self.my_id)
+
+        return shares
+
+    def _get_dealer_msg(self, value):
+        # Sample a random degree-(t,t) bivariate polynomial φ(·,·)
+        # such that each φ(0,k) = sk and φ(i,k) is Pi’s share of sk
+        phi = [None] * (self.t+1)
+        commitments = [None] * (self.t+1)
+        aux_poly = [None] * (self.t+1)
+        # for k ∈ [t+1]
+        #   Ck, auxk <- PolyCommit(SP,φ(·,k))
+        for k in range(self.t+1):
+            phi[k] = self.poly.random(self.t, value[k])
+            commitments[k], aux_poly[k] = self.poly_commit.commit(phi[k])
+
+        # for each party pi and each k ∈ [t+1]
+        #   1. w[i][k] <- CreateWitnesss(Ck,auxk,i)
+        #   2. z[i][k] <- EncPKi(φ(i,k), w[i][k])
+        ephemeral_secret_key = self.field.random()
+        ephemeral_public_key = pow(self.g, ephemeral_secret_key)
+
+        # construct the payload z
+        z_size = self.n * (self.t + 1)
+        z = [None] * z_size
+        for k in range(z_size):
+            i = k // (self.t + 1)
+            j = k % (self.t + 1)
+            shared_key = pow(self.public_keys[i], ephemeral_secret_key)
```

Comment:
Do not recompute this, you can run a loop outside to compute this and the reuse the keys. You need this value per row and you are computing it for every cell.

Reply:
Resolved.